File: OrderManagementService/Models/Coupon.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using System;
using System.Collections.Generic;

namespace OrderManagementService.Models
{
    public class Coupon
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string? Id { get; set; }

        public string Code { get; set; } = null!;

        public decimal DiscountValue { get; set; }

        public int UsageLimit { get; set; } = 10;

        public int UsageCount { get; set; } = 0;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        [BsonRepresentation(BsonType.ObjectId)]
        public List<string> OrderIds { get; set; } = new();
    }
}

File: frontend/lib/models/Coupon.dart
class Coupon {
  final String? id;
  final String code;
  final int discountValue;
  final int usageLimit;
  final int usageCount;
  final DateTime createdAt;
  final List<String> orderIds;

  Coupon({
    this.id,
    required this.code,
    required this.discountValue,
    required this.usageLimit,
    required this.usageCount,
    required this.createdAt,
    required this.orderIds,
  });

  factory Coupon.fromJson(Map<String, dynamic> json) {
    return Coupon(
      id: json['id'],
      code: json['code'],
      discountValue: json['discountValue'],
      usageLimit: json['usageLimit'],
      usageCount: json['usageCount'],
      createdAt: DateTime.parse(json['createdAt']),
      orderIds: List<String>.from(json['orderIds'] ?? []),
    );
  }

  Map<String, dynamic> toJson() {
    final map = {
      'code': code,
      'discountValue': discountValue,
      'usageLimit': usageLimit,
      'usageCount': usageCount,
      'createdAt': createdAt.toIso8601String(),
      'orderIds': orderIds,
    };
    if (id != null && id!.isNotEmpty) {
      map['id'] = id as Object;
    }
    return map;
  }
}
